### PR TITLE
Add support for fancy diagnostics printing

### PR DESCRIPTION
**Pull request by viniciusmuller**
This pull request adds support printing diagnostics with snippets.

We will get the snippets by default and allow callers to provide `snippet: false` in `Code.print_diagnostic`.

For IO.warn, we will get the snippet whenever we get an `__ENV__` or keyword list as arguments (which implies something at compile-time) and print in the standard way if we get a stacktrace (which implies runtime)

I've also had to update a few tests (~10) that were asserting directly on the old diagnostic format, so I guess it's fine for them to be in this PR.

### CODE REVIEW

File: lib/elixir/lib/code.ex [hunk 1]
```diff
@@ -588,11 +588,17 @@ defmodule Code do
 
   A diagnostic is either returned by `Kernel.ParallelCompiler`
   or by `Code.with_diagnostics/2`.
+
+  ## Options
+
+  * `:fancy` - whether to pretty-print the diagnostic. As it may impact
+    performance, it is not recommended to be used in runtime. Defaults to `true`.
```
Comment (josevalim) [inline lib/elixir/lib/code.ex:595]: No option. We will always do it by default. :)

File: lib/elixir/src/elixir_errors.erl [hunk 2]
```diff
@@ -50,7 +62,7 @@ print_diagnostic(#{severity := Severity, message := Message, stacktrace := Stack
   io:put_chars(standard_error, [prefix(Severity), Message, Location, "\n\n"]),
   Diagnostic.
 
-emit_diagnostic(Severity, Position, File, Message, Stacktrace) ->
+emit_diagnostic(Severity, Position, File, Message, Stacktrace, IsFancy) ->
```
Comment (josevalim) [inline lib/elixir/src/elixir_errors.erl:65]: This option is not about being fancy or not but rather, it is an option if we should read the snippet or not.

Comment (josevalim) [reply lib/elixir/src/elixir_errors.erl:65]: If we cannot read the snippet, then we still show it fancy, but without the snippet.

Comment (viniciusmuller) [reply lib/elixir/src/elixir_errors.erl:46]: That makes sense, thanks for the feedback! I've updated the code so now `Code.print_diagnostic` reads the snippet by default and `:elixir_errors.print_diagnostic` always uses `format_snippet` but the caller can decide whether to read the snippet

Comment (josevalim) [reply lib/elixir/src/elixir_errors.erl:46]: We can have an option called snippet: true | false.

File: lib/elixir/lib/code.ex [hunk 3]
```diff
@@ -588,11 +588,18 @@ defmodule Code do
 
   A diagnostic is either returned by `Kernel.ParallelCompiler`
   or by `Code.with_diagnostics/2`.
+
+  ## Options
+
+  * `:snippet` - whether to read the code snippet in the diagnostic location.
```
Comment (josevalim) [inline lib/elixir/lib/code.ex:594]:
```suggestion
    * `:snippet` - whether to read the code snippet in the diagnostic location.
```

File: lib/elixir/src/elixir_errors.erl (hunk 2 above)
Comment (josevalim) [reply lib/elixir/src/elixir_errors.erl:46]: Don't forget to rename this variable. :)

Comment (josevalim) [reply lib/elixir/src/elixir_errors.erl:46]:
```suggestion
emit_diagnostic(Severity, Position, File, Message, Stacktrace, ReadSnippet) ->
```

File: lib/elixir/src/elixir_errors.erl [hunk 4]
```diff
@@ -60,8 +53,8 @@ emit_diagnostic(Severity, Position, File, Message, Stacktrace) ->
   },
 
   case get(elixir_code_diagnostics) of
-    undefined -> print_diagnostic(Diagnostic);
-    {Tail, true} -> put(elixir_code_diagnostics, {[print_diagnostic(Diagnostic) | Tail], true});
+    undefined -> print_diagnostic(Diagnostic, IsFancy);
+    {Tail, true} -> put(elixir_code_diagnostics, {[print_diagnostic(Diagnostic, IsFancy) | Tail], true});
```
Comment (josevalim) [inline lib/elixir/src/elixir_errors.erl:57]:
```suggestion
    undefined -> print_diagnostic(Diagnostic, ReadSnippet);
    {Tail, true} -> put(elixir_code_diagnostics, {[print_diagnostic(Diagnostic, ReadSnippet) | Tail], true});
```

File: lib/elixir/lib/io.ex [hunk 5]
```diff
@@ -329,11 +329,11 @@ defmodule IO do
 
   def warn(message, %Macro.Env{line: line, file: file} = env) do
     message = to_chardata(message)
-    :elixir_errors.emit_diagnostic(:warning, line, file, message, Macro.Env.stacktrace(env))
+    :elixir_errors.emit_diagnostic(:warning, line, file, message, Macro.Env.stacktrace(env), true)
   end
 
   def warn(message, []) do
-    :elixir_errors.emit_diagnostic(:warning, 0, nil, to_chardata(message), [])
+    :elixir_errors.emit_diagnostic(:warning, 0, nil, to_chardata(message), [], true)
```
Comment (josevalim) [inline lib/elixir/lib/io.ex:336]:
```suggestion
    :elixir_errors.emit_diagnostic(:warning, 0, nil, to_chardata(message), [], false)
```

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**